proposals/0012-introduce-reactNativeMeta.md

```diff
@@ -0,0 +1,148 @@
+
+There are a few areas that this section could express in a consistent manner; here's a list of the ones we could think of (in no particular order):
+
+- if a library supports the new architecture, turbo modules or fabric or both
+- which versions of react-native a library supports in a given release
+- if an app has opted in into Fabric, TM, or both
+- if an app is using Hermes or another js engine
+
+Sidenote: the name `reactNativeMeta` was chosen to mirror the camel case pattern used by other fields such as `peerDependenciesMeta`.
+
+## Detailed design
+
+This is the section that in this first draft will need to be properly finalised. For now, here's what we envisioned:
+
+```js
+reactNativeMeta: {
+    version: string, // ex. 1.0.0 - we know that this spec might evolve in the future
+    type: string; // "app" or "library"
+    features: {
+        // can be global...
+        newArchEnabled: boolean,
+        jsEngine: string, // ex. "hermes", "jsc", "V8"...
+        // ...or platform specific
+        android: {
+            newArchEnabled: boolean,
+            fabric: boolean,
+            turbomodules: boolean,
+            jsEngine: string,
+        },
+        iOs: {
+        ... // same as android
+        },
+        macOs: {
+        ... // same as android
+        },
+        windows: {
+        ... // same as android
+        },
+        // more platforms, params - add a comment here
+    },
+    // this section would most likely only be relevant for libraries
+    requirements: {
+        "react-native": string, // semver compliant range of versions of RN supported
+        expo: string, // semver compliant range of versions of Expo supported (if any)
+    }
+}
+```
+
+## Drawbacks
+
+The main drawback is that we'd need to convince developers to start using it, so we'd have to expect a phase in which only a part of the userbase has this section configured and tool developers need to account for the section missing entirely.
+
+## Alternatives
+
+`reactNativeMeta` is, by design, fully original and new: there are other files that are used to define configurations for various aspect of react-native based projects (such as `react-native.config.js`, `app.json`, Expo's `app.config.js`, `expo-module.config.json`) but this one does **not** overlap with any of them. Explicitly, this one has the purpose of being filled only with metadata to be read by package managers and other tools.
+
+## Adoption strategy
+
+Once the v1.0.0 version of the shape is finalised, we can start integrating it across the board in tools such has `align-deps` and the RNDirectory, and in the react-native tools themselves. This first wave of support would have to be coordinated around a certain release that lives in RN's main branch, such as now it'd be introduced in 0.72 (which branch has not been cut).
```

Comment:
For the app config, would the idea be to also teach native build logic how to read the flags?

Reply:
yes this section can be clarified and expanded to also mention how this would show up in the template, I'll do another pass tomorrow to add more info also based on other comments.

Reply:
Clarified on 76cb7db062beb14af3a610a87bce49785f51daa2